### PR TITLE
Add load balancer application profile CRUD

### PR DESCRIPTION
**Pull request by Didainius**
As part of load balancer introduction (ref  https://github.com/terraform-providers/terraform-provider-vcd/issues/223) this PR introduces application profile support in Edge gateway.

**Note** The struct `LBAppProfile` has field `Expire` which is available in GUI, documented in API documentation but seems to be ignored. I'm researching this more. Other than that the PR should be working fine.

### CODE REVIEW

File: govcd/lbservicemonitor.go [hunk 1]
```diff
@@ -33,16 +32,13 @@ func (eGW *EdgeGateway) CreateLBServiceMonitor(lbMonitorConfig *types.LBMonitor)
...
 	}
-	location := resp.Header.Get("Location")
 
-	// Last element in location header is the service monitor ID
-	// i.e. Location: [/network/edges/edge-3/loadbalancer/config/monitors/monitor-5]
-	// The code below extracts that ID from the last segment
-	if location == "" {
-		return nil, fmt.Errorf("unable to retrieve ID for new load balancer service monitor with name %s", lbMonitorConfig.Name)
+	// Location header should look similarly:
+	// Location: [/network/edges/edge-3/loadbalancer/config/monitors/monitor-5]
```
Comment (lvirbalas) [inline govcd/lbservicemonitor.go]: Nit: "Location header should look similar to:"?

Comment (Didainius) [reply govcd/lbservicemonitor.go]: Changed here and in other load balancer resources.

File: govcd/lb.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+package govcd
+
+import (
+	"fmt"
+	"path"
+	"strings"
+)
+
+func extractNSXObjectIDfromPath(locationPath string) (string, error) {
```
Comment (dataclouder) [inline govcd/lb.go]: Some comments on what this function does would be helpful.
Specifically, why do we expect a 2-element list from SplitPath

File: govcd/lbappprofile_test.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	check.Assert(createdLbAppProfile.Template, Equals, lbAppProfileConfig.Template)
+
+	// Test updating fields
+	// Update persistence method
+	lbAppProfileByID.Persistence.Method = "sourceip"
+	updatedAppProfile, err := edge.UpdateLBAppProfile(lbAppProfileByID)
+	check.Assert(err, IsNil)
+	check.Assert(updatedAppProfile.Persistence.Method, Equals, lbAppProfileByID.Persistence.Method)
+
+	// Verify that updated pool and it's configuration are identical
```
Comment (dataclouder) [inline govcd/lbappprofile_test.go]: "it's" -> "its"

File: govcd/lbappprofile.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+		if lbAppProfileConfig.ID != "" && profile.ID == lbAppProfileConfig.ID {
+			return profile, nil
+		}
+
+		// If Name was specified for lookup - look for the same Name
+		if lbAppProfileConfig.Name != "" && profile.Name == lbAppProfileConfig.Name {
+			// We found it by name. Let's verify if search ID was specified and it matches the lookup object
+			if lbAppProfileConfig.ID != "" && profile.ID != lbAppProfileConfig.ID {
+				return nil, fmt.Errorf("load balancer application profile was found by name (%s)"+
+					", but it's ID (%s) does not match specified ID (%s)",
```
Comment (dataclouder) [inline govcd/lbappprofile.go]: "it's" -> "its"

File: govcd/lbappprofile_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	check.Assert(updatedAppProfile, DeepEquals, lbAppProfileByID)
+
+	// Try to set invalid algorithm hash and excpect API to return error
```
Comment (dataclouder) [inline govcd/lbappprofile_test.go]: "excpect" -> "expect"

Comment (Didainius) [reply govcd/lbappprofile_test.go]: Done

File: govcd/lbappprofile.go (hunk 4 above)
Comment (Didainius) [reply govcd/lbappprofile.go]: Done.

File: govcd/lbappprofile_test.go (hunk 3 above)
Comment (Didainius) [reply govcd/lbappprofile_test.go]: Done.

File: govcd/lb.go (hunk 2 above)
Comment (Didainius) [reply govcd/lb.go]: Added

File: govcd/lbappprofile.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	// Location header should look similar to:
+	// [/network/edges/edge-3/loadbalancer/config/applicationprofiles/applicationProfile-4]
+	lbAppProfileID, err := extractNSXObjectIDfromPath(resp.Header.Get("Location"))
+	if err != nil {
+		return nil, err
+	}
+
+	readAppProfile, err := eGW.ReadLBAppProfile(&types.LBAppProfile{ID: lbAppProfileID})
```
Comment (vbauzysvmware) [inline govcd/lbappprofile.go]: We could use `ReadLBAppProfileById`

File: govcd/lbappprofile.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+// If both - Name and ID are specified it performs a lookup by ID and returns an error if the specified name and found
+// name do not match.
+func (eGW *EdgeGateway) UpdateLBAppProfile(lbAppProfileConfig *types.LBAppProfile) (*types.LBAppProfile, error) {
+	if err := validateUpdateLBAppProfile(lbAppProfileConfig); err != nil {
+		return nil, err
+	}
+
+	// if only name was specified for update, ID must be found, because ID is mandatory for update
+	if lbAppProfileConfig.ID == "" {
+		readlbAppProfile, err := eGW.ReadLBAppProfile(&types.LBAppProfile{Name: lbAppProfileConfig.Name})
```
Comment (vbauzysvmware) [inline govcd/lbappprofile.go]: Could use ReadLBAppProfileByName

File: govcd/lbappprofile.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	// Result should be 204, if not we expect an error of type types.NSXError
+	_, err = eGW.client.ExecuteRequestWithCustomError(httpPath, http.MethodPut, types.AnyXMLMime,
+		"error while updating load balancer application profile : %s", lbAppProfileConfig, &types.NSXError{})
+	if err != nil {
+		return nil, err
+	}
+
+	readAppProfile, err := eGW.ReadLBAppProfile(&types.LBAppProfile{ID: lbAppProfileConfig.ID})
```
Comment (vbauzysvmware) [inline govcd/lbappprofile.go]: Could use ReadLBAppProfileById

File: govcd/lbappprofile.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+// name do not match.
+func (eGW *EdgeGateway) DeleteLBAppProfile(lbAppProfileConfig *types.LBAppProfile) error {
+	if err := validateDeleteLBAppProfile(lbAppProfileConfig); err != nil {
+		return err
+	}
+
+	lbAppProfileID := lbAppProfileConfig.ID
+	// if only name was specified for deletion, ID must be found, because only ID can be used for deletion
+	if lbAppProfileConfig.ID == "" {
+		readlbAppProfile, err := eGW.ReadLBAppProfile(&types.LBAppProfile{Name: lbAppProfileConfig.Name})
```
Comment (vbauzysvmware) [inline govcd/lbappprofile.go]: Could you use ReadLBAppProfileByName

File: govcd/lbappprofile.go [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+// DeleteLBAppProfile is able to delete the types.LBAppProfile type by Name and/or ID.
+// If both - Name and ID are specified it performs a lookup by ID and returns an error if the specified name and found
+// name do not match.
+func (eGW *EdgeGateway) DeleteLBAppProfile(lbAppProfileConfig *types.LBAppProfile) error {
+	if err := validateDeleteLBAppProfile(lbAppProfileConfig); err != nil {
+		return err
+	}
+
+	lbAppProfileID := lbAppProfileConfig.ID
```
Comment (vbauzysvmware) [inline govcd/lbappprofile.go]: Seems getting code part duplicates through methods and could be moved to separate method

File: govcd/lbappprofile_test.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+package govcd
+
+import (
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	. "gopkg.in/check.v1"
+)
+
+func (vcd *TestVCD) Test_LBAppProfile(check *C) {
```
Comment (vbauzysvmware) [inline govcd/lbappprofile_test.go]: comment would help here - what is testing

File: types/v56/types.go [hunk 12]
```diff
@@ -1677,6 +1677,41 @@ type LBPoolMember struct {
 
 type LBPoolMembers []LBPoolMember
 
+// LBAppProfile represents a load balancer application profile as per "vCloud Director API for NSX
+// Programming Guide"
+// Type: LBPoolHealthCheckType
```
Comment (vbauzysvmware) [inline types/v56/types.go]: Is this right?

File: govcd/lbappprofile_test.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+	lbAppProfileByID.Persistence.Method = "sourceip"
+	updatedAppProfile, err := edge.UpdateLBAppProfile(lbAppProfileByID)
+	check.Assert(err, IsNil)
+	check.Assert(updatedAppProfile.Persistence.Method, Equals, lbAppProfileByID.Persistence.Method)
+
+	// Verify that updated pool and its configuration are identical
+	check.Assert(updatedAppProfile, DeepEquals, lbAppProfileByID)
+
+	// Try to set invalid algorithm hash and expect API to return error
+	// Invalid persistence method invalid_method. Valid methods are: COOKIE|SSL-SESSIONID|SOURCEIP.
```
Comment (vbauzysvmware) [inline govcd/lbappprofile_test.go]: Is API error self explanatory and we don't need validate methods types?

File: govcd/lbappprofile.go (hunk 6 above)
Comment (Didainius) [reply govcd/lbappprofile.go]: Using.

File: govcd/lbappprofile.go (hunk 7 above)
Comment (Didainius) [reply govcd/lbappprofile.go]: Using.

File: govcd/lbappprofile.go (hunk 8 above)
Comment (Didainius) [reply govcd/lbappprofile.go]: Using.

File: govcd/lbappprofile.go (hunk 9 above)
Comment (Didainius) [reply govcd/lbappprofile.go]: Using.

File: govcd/lbappprofile.go (hunk 10 above)
Comment (Didainius) [reply govcd/lbappprofile.go]: I split it. How does it look now. If it's fine I will retrofit the other components.

File: govcd/lbappprofile_test.go (hunk 11 above)
Comment (Didainius) [reply govcd/lbappprofile_test.go]: Added.

File: types/v56/types.go (hunk 12 above)
Comment (Didainius) [reply types/v56/types.go]: Nope - it's not. remove the `type`

File: govcd/lbappprofile_test.go (hunk 13 above)
Comment (Didainius) [reply govcd/lbappprofile_test.go]: It is self explanatory. In fact this line is part of API response.
`Invalid persistence method invalid_method. Valid methods are: COOKIE|SSL-SESSIONID|SOURCEIP.`

File: govcd/lbappprofile.go (hunk 10 above)
Comment (vbauzysvmware) [reply govcd/lbappprofile.go]: I think better :)

Comment (Didainius) [reply govcd/lbappprofile.go]: Made the same thing for other components.

Comment (vbauzysvmware) [reply govcd/lbappprofile.go]: nice

**Pull request merged**